File: CSharp/Data_Structures/LinkedLists/DoublyLinkedList.cs
namespace CSharp.Data_Structures.LinkedLists;


class DLLNode{
    public Object value;
    public DLLNode? prev;
    public DLLNode? next;

    public DLLNode(Object value){
        this.value = value;
        this.prev = this.next = null;
    }
}

public class DoublyLinkedList
{
    DLLNode head;
    DLLNode tail;
    int length;

    DoublyLinkedList()
    {
        this.head = this.tail = null;
        this.length = 0;
    }

    Object GetFirst()
    {
        if (this.head == null)
        {
            throw new InvalidOperationException();
        }

        return this.head.value;
    }

    Object GetLast()
    {
        if (this.head == null)
        {
            throw new InvalidOperationException();
        }

        return this.tail.value;
    }

    DLLNode Traverse(int index)
    {
        DLLNode currentNode;
        int position;
        if(this.length - index < index){
            currentNode = this.tail;
            position = this.length - 1;
            while(position != index){
                currentNode = currentNode.prev;
                position--;
            }
        }
        else {
            currentNode = this.head;
            position = 0;
            while (position != index) {
                currentNode = currentNode.next;
                position++;
            }
        }
        return currentNode;
    }
    
    Object Get(int index)
    {
        if (this.head == null) {
            throw new InvalidOperationException("List is empty");
        }
        if (index < 0 || index >= this.length) {
            throw new IndexOutOfRangeException("Index out of range");
        }
        DLLNode currentNode = Traverse(index);
        return currentNode.value;
    }
    
    void Append(Object value){
        DLLNode newNode = new DLLNode(value);
        if(this.head == null){
            this.head = this.tail = newNode;
        }
        else{
            DLLNode lastNode = this.tail;
            lastNode.next = newNode;
            newNode.prev = lastNode;
            this.tail = newNode;
        }
        this.length++;
    }
    
    void Prepend(Object value){
        DLLNode newNode = new DLLNode(value);
        if(this.head == null){
            this.head = this.tail = newNode;
        }
        else{
            DLLNode firstNode = this.head;
            firstNode.prev = newNode;
            newNode.next = firstNode;
            this.head = newNode;
        }
        this.length++;
    }
    
    void Insert(int index, Object value){
        if (this.head == null) {
            throw new InvalidOperationException("List is empty");
        }
        if (index < 0 || index > this.length) {
            throw new IndexOutOfRangeException("Index out of range");
        }
        if(index == 0){
            Prepend(value);
        }
        if(index == this.length) {
            Append(value);
        }
        else{
            DLLNode newNode = new DLLNode(value);
            DLLNode previousNode = Traverse(index - 1);
            DLLNode nextNode = previousNode.next;
            newNode.next = nextNode;
            newNode.prev = previousNode;
            previousNode.next = nextNode.prev = newNode;
            this.length++;
        }
    }
    
    void Remove(int index){
        if (this.head == null) {
            throw new InvalidOperationException("List is empty");
        }
        if (index < 0 || index >= this.length) {
            throw new IndexOutOfRangeException("Index out of range");
        }
        if(index == 0){
            this.head = this.head.next;
            this.head.prev = null;
            if(this.length == 1){
                this.tail = null;
            }
        }
        else{
            DLLNode previousNode = Traverse(index - 1);
            DLLNode currentNode = previousNode.next;
            DLLNode nextNode = currentNode.next;
            previousNode.next = nextNode;
            if (nextNode != null){
                nextNode.prev = previousNode;
            }
            else{
                this.tail = previousNode;
            }
        }
        this.length--;
    }
    
    void Display(){
        DLLNode currentNode = this.head!;
        while(currentNode != null){
            Console.Write(currentNode.value + " <-> ");
            currentNode = currentNode.next;
        }
        Console.Write("null");
        Console.WriteLine();
    }
    
    public static void Main(String[] args) {
        DoublyLinkedList myDoublyList = new DoublyLinkedList();
        myDoublyList.Display();
        myDoublyList.Append(1);
        myDoublyList.Append(9);
        myDoublyList.Append(5);
        myDoublyList.Prepend(4);
        myDoublyList.Insert(1, 50);
        Console.WriteLine(myDoublyList.Get(2));
        myDoublyList.Display();
        myDoublyList.Remove(4);
        myDoublyList.Display();
        Console.WriteLine(myDoublyList.GetFirst());
        Console.WriteLine(myDoublyList.GetLast());
    }
}

File: Python/Data_Structures/LinkedLists/DoublyLinkedList.py
class DLLNode:
    def __init__(self, value):
        self.value = value
        self.prev = self.next = None


class MyDoublyLinkedList:
    def __init__(self):
        self.head = self.tail = None
        self.length = 0

    def get_first(self):
        if self.head is None:
            raise Exception("List is empty")
        return self.head.value

    def get_last(self):
        if self.head is None:
            raise Exception("List is empty")
        return self.tail.value

    def traverse(self, index):
        if self.length - index < index:
            current_node = self.tail
            position = self.length - 1
            while position != index:
                current_node = current_node.prev
                position -= 1
        else:
            current_node = self.head
            position = 0
            while position != index:
                current_node = current_node.next
                position += 1
        return current_node

    def get(self, index):
        if index < 0 or index >= self.length:
            raise IndexError("Index out of range")
        if self.head is None:
            raise Exception("List is empty")
        return self.traverse(index).value

    def append(self, value):
        new_node = DLLNode(value)
        if self.head is None:
            self.head = self.tail = new_node
        else:
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = new_node
        self.length += 1

    def prepend(self, value):
        new_node = DLLNode(value)
        if self.head is None:
            self.head = self.tail = new_node
        else:
            self.head.prev = new_node
            new_node.next = self.head
            self.head = new_node
        self.length += 1

    def insert(self, index, value):
        if self.head is None:
            raise Exception("List is empty")
        if index < 0 or index > self.length:
            raise IndexError("Index out of range")
        if index == 0:
            self.prepend(value)
        if index == self.length:
            self.append(value)
        else:
            new_node = DLLNode(value)
            previous_node = self.traverse(index - 1)
            next_node = previous_node.next
            new_node.next = next_node
            new_node.prev = previous_node
            previous_node.next = next_node.prev = new_node
            self.length += 1

    def remove(self, index):
        if self.head is None:
            raise Exception("List is empty")
        if index < 0 or index >= self.length:
            raise IndexError("Index out of range")
        if index == 0:
            self.head = self.head.next
            if self.head is not None:
                self.head.prev = None
            if self.length == 1:
                self.tail = None
        else:
            previous_node = self.traverse(index - 1)
            current_node = previous_node.next
            next_node = current_node.next
            previous_node.next = next_node
            if next_node is not None:
                next_node.prev = previous_node
            else:
                self.tail = previous_node
        self.length -= 1

    def display(self):
        current_node = self.head
        output = ''
        while current_node is not None:
            output += f'{current_node.value} <-> '
            current_node = current_node.next
        output += 'null'
        print(output)


if __name__ == '__main__':
    myDoublyList = MyDoublyLinkedList()
    myDoublyList.display()
    myDoublyList.append(3)
    myDoublyList.append(7)
    myDoublyList.append(4)
    myDoublyList.prepend(5)
    myDoublyList.insert(1, 45)
    myDoublyList.display()
    myDoublyList.remove(4)
    myDoublyList.display()
    print(myDoublyList.get_first())
    print(myDoublyList.get_last())

